shell_integration/MacOSX/OwnCloudFinderSync/FinderSyncExt/FinderSync.m

```diff
@@ -101,54 +115,43 @@ - (NSMenu *)menuForMenuKind:(FIMenuKind)whichMenu
-        NSMenu *menu = [[NSMenu alloc] initWithTitle:@""];
-        NSMenu *subMenu = [[NSMenu alloc] initWithTitle:@""];
-        NSMenuItem *subMenuItem = [menu addItemWithTitle:contextMenuTitle action:nil keyEquivalent:@""];
-        subMenuItem.submenu = subMenu;
-        subMenuItem.image = [[NSBundle mainBundle] imageForResource:@"app.icns"];
-
-		[subMenu addItemWithTitle:shareTitle action:@selector(shareMenuAction:) keyEquivalent:@""];
-		[subMenu addItemWithTitle:copyLinkTitle action:@selector(copyLinkMenuAction:) keyEquivalent:@""];
-		[subMenu addItemWithTitle:emailLinkTitle action:@selector(emailLinkMenuAction:) keyEquivalent:@""];
-
+		NSMenu *menu = [[NSMenu alloc] initWithTitle:@""];
+		NSMenu *subMenu = [[NSMenu alloc] initWithTitle:@""];
+		NSMenuItem *subMenuItem = [menu addItemWithTitle:contextMenuTitle action:nil keyEquivalent:@""];
+		subMenuItem.submenu = subMenu;
+		subMenuItem.image = [[NSBundle mainBundle] imageForResource:@"app.icns"];
+
+		// There is an annoying bug in macOS (at least 10.13.3), it does not use/copy over the representedObject of a menu item
+		// So we have to use tag instead.
+		int idx = 0;
+		for (NSArray* item in _menuItems) {
+			NSMenuItem *actionItem = [subMenu addItemWithTitle:[item valueForKey:@"text"]
+														action:@selector(subMenuActionClicked:)
+												 keyEquivalent:@""];
+			[actionItem setTag:idx];
+			[actionItem setTarget:self];
+			NSString *flags = [item valueForKey:@"flags"]; // e.g. "d"
+			if ([flags rangeOfString:@"d"].location != NSNotFound) {
+				[actionItem setEnabled:false];
+			}
+			idx++;
+		}
 		return menu;
 	}
 	return nil;
 }
 
-- (IBAction)shareMenuAction:(id)sender
-{
-	NSArray* items = [[FIFinderSyncController defaultController] selectedItemURLs];
-
-	[items enumerateObjectsUsingBlock: ^(id obj, NSUInteger idx, BOOL *stop) {
-		NSString* normalizedPath = [[obj path] decomposedStringWithCanonicalMapping];
-		[_syncClientProxy askOnSocket:normalizedPath query:@"SHARE"];
-	}];
-}
-
-- (IBAction)copyLinkMenuAction:(id)sender
-{
-	NSArray* items = [[FIFinderSyncController defaultController] selectedItemURLs];
-
-	[items enumerateObjectsUsingBlock: ^(id obj, NSUInteger idx, BOOL *stop) {
-		NSString* normalizedPath = [[obj path] decomposedStringWithCanonicalMapping];
-		[_syncClientProxy askOnSocket:normalizedPath query:@"COPY_PRIVATE_LINK"];
-	}];
-}
-
-- (IBAction)emailLinkMenuAction:(id)sender
-{
-	NSArray* items = [[FIFinderSyncController defaultController] selectedItemURLs];
-
-	[items enumerateObjectsUsingBlock: ^(id obj, NSUInteger idx, BOOL *stop) {
-		NSString* normalizedPath = [[obj path] decomposedStringWithCanonicalMapping];
-		[_syncClientProxy askOnSocket:normalizedPath query:@"EMAIL_PRIVATE_LINK"];
-	}];
+- (void)subMenuActionClicked:(id)sender {
+	long idx = [(NSMenuItem*)sender tag];
+	NSString *command = [[_menuItems objectAtIndex:idx] valueForKey:@"command"];
+	NSString *paths = [self selectedPathsSeparatedByRecordSeparator];
```

Comment:
minor: I hope this returns the same list of paths every time / users can't somehow change selected paths between the two calls